tests/child.cpp

```diff
@@ -0,0 +1,190 @@
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include <sstream>
+#include <string>
+
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <time.h>
+
+#include "child.h"
+#include "utils.h"
+
+namespace bpftrace {
+namespace test {
+namespace child {
+
+using ::testing::HasSubstr;
+
+#define TEST_BIN "/bin/ls"
+#define TEST_BIN_LONG "/bin/sleep 10"
+
+int msleep(int msec)
+{
+  struct timespec sleep = { .tv_sec = 0, .tv_nsec = msec * 1000000L };
+  struct timespec rem = {};
+  if (nanosleep(&sleep, &rem) < 0)
+    return 1000L * rem.tv_sec + 1000000L * rem.tv_nsec;
+  return 0;
+}
+
+void wait_for(ChildProcBase *child, int msec_timeout)
+{
+  constexpr int wait = 10;
+  while (child->is_alive() && msec_timeout > 0)
+    msec_timeout -= wait - msleep(wait);
+}
+
+std::unique_ptr<ChildProc> getChild(std::string cmd)
+{
+  std::unique_ptr<ChildProc> child;
+  {
+    StderrSilencer es;
+    StdoutSilencer os;
+    os.silence();
+    es.silence();
+    child = std::make_unique<ChildProc>(cmd);
+  }
+  EXPECT_NE(child->pid(), -1);
+  EXPECT_TRUE(child->is_alive());
+  return child;
+}
+
+TEST(childproc, exe_does_not_exist)
+{
+  try
+  {
+    auto child = ChildProc("/does/not/exist/abc/fed");
+    FAIL();
+  }
+  catch (const std::runtime_error &e)
+  {
+    EXPECT_THAT(e.what(), HasSubstr("does not exist or is not executable"));
+  }
+}
+
+TEST(childproc, too_many_arguments)
+{
+  std::stringstream cmd;
+  cmd << "/bin/ls";
+  for (int i = 0; i < 280; i++)
+    cmd << " a";
+
+  try
+  {
+    auto child = ChildProc(cmd.str());
+    FAIL();
+  }
+  catch (const std::runtime_error &e)
+  {
+    EXPECT_THAT(e.what(), HasSubstr("Too many arguments"));
+  }
+}
+
+TEST(childproc, child_exit_success)
+{
+  // Spawn a child that exits  success
+  auto child = getChild(TEST_BIN);
+
+  child->run();
+  wait_for(child.get(), 1000);
+  EXPECT_FALSE(child->is_alive());
+  EXPECT_EQ(child->exit_code(), 0);
+  EXPECT_EQ(child->term_signal(), -1);
+}
+
+TEST(childproc, child_exit_err)
+{
+  // Spawn a child that exits  success
+  auto child = getChild("/bin/ls /does/not/exist/abc/fed");
+
+  child->run();
+  wait_for(child.get(), 1000);
+  EXPECT_FALSE(child->is_alive());
+  EXPECT_EQ(child->exit_code(), 2);
+  EXPECT_EQ(child->term_signal(), -1);
+}
+
+TEST(childproc, terminate)
+{
+  auto child = getChild(TEST_BIN_LONG);
+
+  child->run();
+  child->terminate();
+  wait_for(child.get(), 100);
+  EXPECT_FALSE(child->is_alive());
+  EXPECT_EQ(child->term_signal(), SIGTERM);
+}
+
+TEST(childproc, destructor_destroy_child)
+{
+  pid_t child_pid = 0;
+  {
+    std::unique_ptr<ChildProc> child = getChild(TEST_BIN_LONG);
+    child->run();
+    child_pid = child->pid();
+    msleep(25);
```

Comment:
This was mainly to get the childproc to actually exec the binary we want instead of killing it straight away. I've added a comment but I think we can remove this too as execve should not matter for this test